python/metric_pipeline_tasks/MatchedCatalogMeasureTasks.py

```diff
@@ -48,6 +47,83 @@ def run(self, matchedCatalog, metric_name):
+    brightSnrMin = Field(doc="Minimum median SNR for a source to be considered bright.",
+                         dtype=float, default=50)
+    brightSnrMax = Field(doc="Maximum median SNR for a source to be considered bright.",
+                         dtype=float, default=np.Inf)
+    threshPA2 = Field(doc="Threshold in mmag for PF1 calculation.", dtype=float, default=15.0)
+    threshPF1 = Field(doc="Percentile of differences that can vary by more than threshPA2.",
+                      dtype=float, default=10.0)
+
+
+class PA2Task(Task):
+
+    ConfigClass = PA2TaskConfig
+    _DefaultName = "PA2Task"
+
+    def __init__(self, config: PA2TaskConfig, *args, **kwargs):
+        super().__init__(*args, config=config, **kwargs)
+        self.brightSnrMin = self.config.brightSnrMin
+        self.brightSnrMax = self.config.brightSnrMax
+        self.threshPA2 = self.config.threshPA2
+        self.threshPF1 = self.config.threshPF1
+
+    def run(self, matchedCatalog, metric_name):
+        self.log.info(f"Measuring PA2")
+        pf1_thresh = self.threshPF1 * u.percent
+
+        filteredCat = filterMatches(matchedCatalog) #, extended=False, isPrimary=False)
+        magKey = filteredCat.schema.find('slot_PsfFlux_mag').key
+
+        # Require at least nMinPA2=10 objects to calculate the repeatability:
+        nMinPA2 = 50
+        if filteredCat.count > nMinPA2:
+            pa2 = calcPhotRepeat(filteredCat, magKey)
+            # Previously, validate_drp used the first random sample from PA1 measurement
+            # Now, use all of them.
+            magDiffs = pa2['magDiff']
+
+            pf1Percentile = 100.*u.percent - pf1_thresh
+            return Struct(measurement=Measurement("PA2", np.percentile(np.abs(magDiffs.value),
+                          pf1Percentile.value) * magDiffs.unit))
+        else:
+            return Struct(measurement=Measurement("PA2", np.nan*u.mmag))
+
+
+class PF1Task(Task):
+
+    ConfigClass = PA2TaskConfig
+    _DefaultName = "PF1Task"
+
+    def __init__(self, config: PA2TaskConfig, *args, **kwargs):
+        super().__init__(*args, config=config, **kwargs)
+        self.brightSnrMin = self.config.brightSnrMin
+        self.brightSnrMax = self.config.brightSnrMax
+        self.threshPA2 = self.config.threshPA2
+        self.threshPF1 = self.config.threshPF1
+
+    def run(self, matchedCatalog, metric_name):
+        self.log.info(f"Measuring PF1")
+        pa2_thresh = self.threshPA2 * u.mmag
+
+        filteredCat = filterMatches(matchedCatalog) #, extended=False, isPrimary=False)
+        magKey = filteredCat.schema.find('slot_PsfFlux_mag').key
+
+        # Require at least nMinPA2=10 objects to calculate the repeatability:
+        nMinPA2 = 50
+        if filteredCat.count > nMinPA2:
+            pf1 = calcPhotRepeat(filteredCat, magKey)
+            # Previously, validate_drp used the first random sample from PA1 measurement
+            # Now, use all of them.
+            magDiffs = pf1['magDiff']
```

Comment:
Since the run in PA2 and PF1 (and I suspect PA1) share several lines of identical code, it would be nice to factor that out into a utility function that could be called by both methods.  Especially since there are things like the minimum number of objects to use that would need to be updated in all three places if we decided to change them.
This sort of duplication makes maintenance a headache.

Reply:
Excellent point. I've added a separate function called `photRepeat`, which in turn calls `calcPhotRepeat`. This function is then called by each of PA1, PA2, and PF1 (with the option of passing filtering kwargs).